accepted/2020/or-greater-defines/or-greater-defines.md

```diff
@@ -0,0 +1,218 @@
+            throw new PlatformNotSupportedException();
+        #else
+            #error Provide implementation
+        #endif
+    }
+}
+```
+
+### When running on a later version of .NET, use a newer API
+
+Zachariah is building a .NET imaging library. On .NET 5, we have added enough
+hardware intrinsics so that he can implement the blur functionality in a much
+faster way. However, he still wants to support .NET Standard 2.0, so he decides
+to use `#if` to multi-target for .NET 5.
+
+Since he checks for the minimum version of the framework that introduced the
+API, adding a new version wouldn't change this code, so he simply puts the
+fallback logic for the slow version in the `#else` branch, with no call to
+`#error`.
+
+```C#
+public class Picture
+{
+    public void Blur(float radius)
+    {
+        #if NET5_0_OR_HIGHER
+            FastBlurUsingHardwareIntrinsics();
+        #else
+            SlowBlurUsingSimpleMath();
+        #endif
+    }
+}
+```
+
+## Requirements
+
+### Goals
+
+* The new symbols don't change semantics of existing code
+* The new symbols allow expressing a `>=` relationship with version numbers
+* The new symbols apply to both frameworks as well as platforms
+* The new symbols are defined for all frameworks that are relevant for package
+  authors today, which includes:
+    - .NET 5
+    - .NET Core
+    - .NET Framework
+    - .NET Standard
+
+### Non-Goals
+
+* Making existing code more resilient to TFM changes
+
+## Design
+
+The SDK will define `XXX_OR_GREATER` variants for the following TFMs:
+
+* **.NET Framework**
+    - Applies to `net1.0`-`net4x` and the `NETFRAMEWORK`/`NET` symbols
+    - For example, `net4.8` will define `NETFRAMEWORK`, `NET48`,
+      `NET48_OR_GREATER` `NET472_OR_GREATER`, `NET471_OR_GREATER`, etc.
+* .NET Core
+    - Applies to `netcoreappX.Y` and the `NETCOREAPP` symbol
+    - For example, `netcoreapp3.1` will define `NETCOREAPP`, `NETCOREAPP3_1`,
+      `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER` etc.
+* .NET 5 and later
+    - Applies to `netX.Y` and the `NET` symbol
+    - For example, `net6.0` will define `NET`, `NET6_0`, `NET6_0_OR_GREATER`,
+      `NET5_0_OR_GREATER` etc.
+    - This will also include the corresponding defines a .NET Core 3.1 successor
+      would have gotten. For example, `net5.0` will also define `NETCOREAPP`,
+      `NETCOREAPP3_1_OR_GREATER`, `NETCOREAPP3_0_OR_GREATER` etc. But it will
+      *neither* define a `NETCOREAPP5_0` or `NETCOREAPP5_0_OR_GREATER`.
+* .NET 5 and later with operating systems
+    - The OS flavors will also gain the `XXX_OR_GREATER` variants
+    - For example, `net5.0-windows10.0.19222.0` will also define `WINDOWS`,
+      `WINDOWS10_0_19041_0`, `WINDOWS10_0_19041_0_OR_GREATER`, etc.
+* Xamarin
+    - This covers the existing Xamarin offerings, the new .NET 6-based Xamarin
+      is handled by the previous sections.
+    - The existing Xamarin platforms don't have versioned preprocessor symbols
```

Comment:
I think there is some level of preprocessor symbols defined for Android API levels.  @mhutch @jonathanpeppers, @Redth

Reply:
Here is where we generate preprocessor symbols on Android:

https://github.com/xamarin/xamarin-android/blob/5fc88a1a7cf67f2045e790b30e43dd7a5a28cd4f/src/Xamarin.Android.Build.Tasks/Tasks/GetAndroidDefineConstants.cs#L28-L38

`__ANDROID_16__` will be present for 16 or higher, for example.

We were planning on keeping these for .NET 6 in addition to any new ones, so we don't break existing code.

Reply:
I'll add this to the sepc

Reply:
@Jonathanpeppers @mhutch 

I've spec'ed this as follows. Please let me know if you have any concerns:

> * Xamarin
>    - This covers the existing Xamarin offerings, the new .NET 6-based iOS and
>      Android support is handled by the previous sections.
>    - The existing Xamarin Apple platforms don't have versioned preprocessor
>      symbols, but Android does: `__ANDROID_16__`
>    - We don't plan on adding `OR_GREATER` flavors like
>      `__ANDROID_16_OR_GREATER__`. Instead, we're only doing those for .NET 6.